evap/results/views.py

```diff
@@ -1,59 +1,83 @@
+    courses = get_courses_with_prefetched_data(courses)
+    for course in courses:
+        assert course.state == 'published'
+        get_template('results_index_course.html').render(dict(course=course, can_user_see_results_page=True))
+        get_template('results_index_course.html').render(dict(course=course, can_user_see_results_page=False))
+        assert get_course_result_template_fragment_cache_key(course.id, True) in caches['results']
+        assert get_course_result_template_fragment_cache_key(course.id, False) in caches['results']
+
+
+def get_courses_with_prefetched_data(courses):
+    if isinstance(courses, QuerySet):
+        courses = (courses
+            .annotate(num_participants=Count("participants", distinct=True), num_voters=Count("voters", distinct=True))
+            .select_related("type")
+            .prefetch_related(
+                "degrees",
+                "semester",
+                Prefetch("contributions", queryset=Contribution.objects.filter(responsible=True).select_related("contributor"), to_attr="responsible_contributions")
+            )
+        )
+        for course in courses:
+            course.responsible_contributors = [contribution.contributor for contribution in course.responsible_contributions]
+    for course in courses:
+        if not course.is_single_result:
+            course.distribution = calculate_average_distribution(course)
+            course.avg_grade = distribution_to_grade(course.distribution)
+        else:
+            course.single_result_rating_result = get_single_result_rating_result(course)
+    return courses
 
-    courses = semester.course_set.filter(state__in=visible_states).prefetch_related("degrees")
 
+@internal_required
+def index(request):
+    semesters = Semester.get_all_with_published_unarchived_results()
+    courses = Course.objects.filter(semester__in=semesters, state='published')
     courses = [course for course in courses if course.can_user_see_course(request.user)]
 
-    for course in courses:
-        course.distribution = calculate_average_distribution(course)
-        course.avg_grade = distribution_to_grade(course.distribution)
-
-    CourseTuple = namedtuple('CourseTuple', ('courses', 'single_results'))
-
-    courses_by_degree = OrderedDict()
-    for degree in Degree.objects.all():
-        courses_by_degree[degree] = CourseTuple([], [])
-    for course in courses:
-        if course.is_single_result:
-            for degree in course.degrees.all():
-                question_result = collect_results(course).questionnaire_results[0].question_results[0]
-                courses_by_degree[degree].single_results.append((course, question_result))
-        else:
-            for degree in course.degrees.all():
-                courses_by_degree[degree].courses.append(course)
+    if request.user.is_reviewer:
+        additional_courses = Course.objects.filter(semester__in=semesters, state__in=['in_evaluation', 'evaluated', 'reviewed'])
```

Comment:
Huh? Doesn't course.can_user_see_course return true in these cases if the user is a reviewer?

Reply:
yes, but above we're only looking for published courses